client/src/includes/initTooltips.test.js

```diff
@@ -0,0 +1,65 @@
+import { initTooltips, initModernDropdown } from './initTooltips';
```

Comment:
With unit tests like this, our goal is just to check the basics and not go too deep into the implementation of `Tippy` and other libraries (we can assume they are unit tested).

What we do want to test is that Tippy gets called when the correct data attributes exist.

Later, not now, we may want to add some more robust unit tests for the esc handling or similar, but the goal of this issue is to just add basic unit tests.

Below is a basic functional suite of unit tests for this code, here are some pointers;

* Start small, I am sure you did start small but it looks like you ran into the issue of the event listener being called when not expected. It would have been good to stop there and rethink the approach, the event listener spy works for other kinds of unit tests but not this.
* Instead we can use `jest.spyOn` to listen to when `tippy` is called (note the syntax is a bit confusing here and can trip you up, we are importing the entire tippy module as `* as tippy` and then adding the spy to `tippy.default`.
* Now we can reliably check that when we call the function we made `initTooltips` it in turn calls Tippy with some kind of params (the data attribute we use AND some options).
* I have also split the two functions into two `describe` blocks as it reads better in the unit test output.
* For the `initModernDropdown` we want to check that it does NOT get called if the data attribute is not present and then we want to check it DOES get called when present. Here again we just want to check that tippy gets called with something that makes sense & references to the correct DOM elements.
* If we need in the future we can extend these tests but from here we will have a good starting point.

I know that giving you the whole unit test file is a bit unhelpful but I could see where you were heading in the code but it was going to lead to more complex unit tests that were not really checking the bits we care about. We care that Tippy is used and used with the right params, that's pretty much it for now.

Please have a full read of this test file, feel free to tweak it and add to it, then apply to your code and ping for another review.

Thanks so much for your efforts here, this is awesome work, writing unit tests is a bit tricky and sorry if the original issue was not as clear as it could have been.

```
import * as tippy from 'tippy.js';

import { initModernDropdown, initTooltips } from './initTooltips';

jest.spyOn(tippy, 'default');

beforeEach(() => {
  jest.clearAllMocks();
});

describe('initTooltips', () => {
  it('should call the Tippy util with the [data-tippy-content] attribute', () => {
    expect(tippy.default).not.toHaveBeenCalled();
    initTooltips();
    expect(tippy.default).toHaveBeenCalledWith('[data-tippy-content]', {
      plugins: [expect.objectContaining({ name: 'hideOnEsc' })],
    });
  });
});

describe('initModernDropdown', () => {
  it('should not call Tippy if there is no element with [data-button-with-dropdown]', () => {
    expect(tippy.default).not.toHaveBeenCalled();

    initModernDropdown();

    expect(tippy.default).not.toHaveBeenCalled();
  });

  it('should call the Tippy util with the [data-button-with-dropdown-toggle] attribute', () => {
    const html = `
    <div data-button-with-dropdown>
      <button id="button" data-button-with-dropdown-toggle>...</button>
      <div id="content" data-button-with-dropdown-content>
        Content
      </div>
    </div>`;
    document.body.innerHTML = html;

    const content = document.getElementById('content');

    expect(tippy.default).not.toHaveBeenCalled();

    initModernDropdown();

    expect(tippy.default).toHaveBeenLastCalledWith(
      document.getElementById('button'),
      expect.objectContaining({
        content,
        trigger: 'click',
        interactive: true,
        theme: 'dropdown',
        placement: 'bottom',
      }),
    );
  });
});

```

Reply:
Thank you so much for this and for your patience. I have gone through it and it makes sense. I think the PR is ready now. Also if there are other tasks like this pls point them my way. I'd like to keep contributing where possible

Reply:
thanks, I will take another look at this PR